cmd/lhsm-plugin-az-core/archive.go

```diff
@@ -17,101 +18,112 @@ type ArchiveOptions struct {
 	AccountName   string
 	ContainerName string
 	ResourceSAS   string
+	MountRoot     string
 	BlobName      string
 	SourcePath    string
 	Credential    azblob.Credential
 	Parallelism   uint16
 	BlockSize     int64
 	Pacer         util.Pacer
 	ExportPrefix  string
+	HNSEnabled    bool
 }
 
-// persist a blob to the local filesystem
-func Archive(o ArchiveOptions) (int64, error) {
+var blobEndPoint string = "https://%s.blob.core.windows.net/"
+var dfsEndPoint string = "https://%s.dfs.core.windows.net/"
+
+//Archive copies local file to HNS
+func Archive(o ArchiveOptions) (size int64, _ error) {
+	util.Log(pipeline.LogInfo, fmt.Sprintf("Archiving %s", o.BlobName))
 	archiveCtx := context.Background()
 	ctx, cancel := context.WithCancel(archiveCtx)
 	defer cancel()
 
 	p := util.NewPipeline(ctx, o.Credential, o.Pacer, azblob.PipelineOptions{})
 
-	//Get the blob URL
-	cURL, _ := url.Parse(fmt.Sprintf("https://%s.blob.core.windows.net/%s%s", o.AccountName, o.ContainerName, o.ResourceSAS))
+	cURL, _ := url.Parse(fmt.Sprintf(blobEndPoint+"%s%s", o.AccountName, o.ContainerName, o.ResourceSAS))
 	containerURL := azblob.NewContainerURL(*cURL, p)
-	blobName := path.Join(o.ExportPrefix, o.BlobName)
-	blobURL := containerURL.NewBlockBlobURL(blobName)
-
-	util.Log(pipeline.LogInfo, fmt.Sprintf("Archiving %s to %s.", o.BlobName, blobURL.String()))
-
-	// open the file to read from
-	file, _ := os.Open(o.SourcePath)
-	fileInfo, _ := file.Stat()
-	defer file.Close()
-
-	//Save owner, perm, group and acl info
-	total := fileInfo.Size()
-	meta := azblob.Metadata{}
-	owner := fmt.Sprintf("%d", fileInfo.Sys().(*syscall.Stat_t).Uid)
-	permissions := fmt.Sprintf("%o", fileInfo.Mode())
-	group := fmt.Sprintf("%d", fileInfo.Sys().(*syscall.Stat_t).Gid)
-	var acl string
-	modTime := fileInfo.ModTime().Format("2006-01-02 15:04:05 -0700")
-
-	resp, err := blobURL.GetAccountInfo(context.Background())
-	if err != nil {
-		util.Log(pipeline.LogError, fmt.Sprintf("Archiving %s. Failed to account info: %s", o.BlobName, err.Error()))
-		return 0, err
-	}
 
-	hnsEnabledAccount := resp.Response().Header.Get("X-Ms-Is-Hns-Enabled") == "true"
-	if hnsEnabledAccount {
-		u, _ := url.Parse(fmt.Sprintf("https://%s.dfs.core.windows.net/%s/%s", o.AccountName, o.ContainerName, blobName))
-		dfsURL := azblob.NewBlockBlobURL(*u, p)
-		aclResp, err := dfsURL.GetAccessControl(ctx, nil, nil, nil, nil, nil, nil, nil, nil)
+	parents := strings.Split(o.BlobName, string(os.PathSeparator))
+	blobPath := o.MountRoot // Initialize the parent starting from root.
+	for _, currDir := range parents {
```

Comment:
We should perhaps do the parent dir property persistence in parallel to uploading the actual file. If we first do all the parent directories and the path is very "deep", then we'll waste quite a lot of time before transferring the actual data and this will impact the tool's performance negatively.